core/trino-main/src/main/java/io/trino/sql/planner/optimizations/DeterminePartitionCount.java

```diff
@@ -0,0 +1,247 @@
+ * used for remote exchanges. It helps to increase the concurrency of the engine in the case of large cluster.
+ * This rule is also cautious about lack of or incorrect statistics therefore it skips for input multiplying nodes like
+ * CROSS JOIN or UNNEST.
+ *
+ * E.g. 1:
+ * Given query: SELECT count(column_a) FROM table_with_stats_a
+ * config:
+ *  MIN_INPUT_ROWS_PER_TASK: 500
+ * Input table row count: 1000
+ * Estimated partition count: Input table row count / MIN_INPUT_ROWS_PER_TASK => 2
+ *
+ * E.g. 2:
+ * Given query: SELECT * FROM table_with_stats_a as a JOIN table_with_stats_b as b ON a.column_b = b.column_b
+ * config:
+ *  MIN_INPUT_ROWS_PER_TASK: 500
+ * Input tables row count: 1000
+ * Join output row count: 5000
+ * Estimated partition count: max((Input table row count / MIN_INPUT_ROWS_PER_TASK), (Join output row count / MIN_INPUT_ROWS_PER_TASK))  => 10
+ */
+public class DeterminePartitionCount
+        implements PlanOptimizer
+{
+    private static final Logger log = Logger.get(DeterminePartitionCount.class);
+    private static final List<Class<? extends PlanNode>> SKIP_PLAN_NODES = ImmutableList.of(TableExecuteNode.class, TableWriterNode.class, MergeWriterNode.class);
+
+    private final StatsCalculator statsCalculator;
+
+    public DeterminePartitionCount(StatsCalculator statsCalculator)
+    {
+        this.statsCalculator = requireNonNull(statsCalculator, "statsCalculator is null");
+    }
+
+    @Override
+    public PlanNode optimize(
+            PlanNode plan,
+            Session session,
+            TypeProvider types,
+            SymbolAllocator symbolAllocator,
+            PlanNodeIdAllocator idAllocator,
+            WarningCollector warningCollector,
+            TableStatsProvider tableStatsProvider)
+    {
+        requireNonNull(plan, "plan is null");
+        requireNonNull(session, "session is null");
+        requireNonNull(types, "types is null");
+        requireNonNull(tableStatsProvider, "tableStatsProvider is null");
+
+        // Skip for write nodes since writing partitioned data with small amount of nodes could cause
+        // memory related issues even when the amount of data is small. Additionally, skip for FTE mode since we
+        // are not using estimated partitionCount in FTE scheduler.
+        if (PlanNodeSearcher.searchFrom(plan).whereIsInstanceOfAny(SKIP_PLAN_NODES).matches()
+                || getRetryPolicy(session) == RetryPolicy.TASK) {
+            return plan;
+        }
+
+        try {
+            return determinePartitionCount(plan, session, types, tableStatsProvider)
+                    .map(partitionCount -> rewriteWith(new Rewriter(partitionCount), plan))
+                    .orElse(plan);
+        }
+        catch (RuntimeException e) {
+            log.warn(e, "Error occurred when determining hash partition count for query %s", session.getQueryId());
```

Comment:
Is it expected to fail?

Reply:
take a look at this thread: https://github.com/trinodb/trino/pull/15489#discussion_r1060735807